File: apps/api/providers/twilio/fake.py
"""Deterministic, in-process communications provider for local development/tests."""

from __future__ import annotations

import hashlib

from providers.communications import PhoneLookup, PhoneVerification, ProviderCall, ProviderMessage


class FakeTwilioProvider:
    """No-network provider; code ``000000`` is the only approved fake code."""

    async def create_verification(self, *, to_number: str, channel: str) -> PhoneVerification:
        return PhoneVerification(
            provider="fake",
            provider_sid=self._sid("verify", to_number, channel),
            status="pending",
            to_number=to_number,
            channel=channel,
        )

    async def approve_verification(self, *, to_number: str, code: str) -> PhoneVerification:
        approved = code == "000000"
        return PhoneVerification(
            provider="fake",
            provider_sid=self._sid("check", to_number),
            status="approved" if approved else "pending",
            to_number=to_number,
            channel="",
            valid=approved,
        )

    async def create_lookup(self, *, number: str, include_line_type: bool = False) -> PhoneLookup:
        return PhoneLookup(provider="fake", number=number, valid=True)

    async def create_message(
        self, *, to_number: str, body: str | None, channel: str, content_sid: str | None = None,
        status_callback: str | None = None,
    ) -> ProviderMessage:
        return ProviderMessage("fake", self._sid("message", to_number, channel), "queued", to_number)

    async def retrieve_message(self, *, provider_sid: str) -> ProviderMessage:
        return ProviderMessage("fake", provider_sid, "queued", "")

    async def create_call(self, *, to_number: str, twiml_url: str) -> ProviderCall:
        return ProviderCall("fake", self._sid("call", to_number), "queued", to_number)

    async def retrieve_call(self, *, provider_sid: str) -> ProviderCall:
        return ProviderCall("fake", provider_sid, "queued", "")

    @staticmethod
    def _sid(*parts: str) -> str:
        return "fake_" + hashlib.sha256("|".join(parts).encode()).hexdigest()[:24]

File: apps/api/src/providers/twilio/fake.ts
/**
 * Deterministic, in-process communications provider for local development and tests.
 *
 * No network calls are made. The magic code `000000` is the only code that
 * approves a fake verification — everything else returns `pending`.
 */

import { createHash } from 'node:crypto'

import {
  EMPTY_VERIFICATION_FIELDS,
  type MessagingProvider,
  type PhoneLookup,
  type PhoneLookupProvider,
  type PhoneVerification,
  type PhoneVerificationProvider,
  type ProviderCall,
  type ProviderMessage,
  type VoiceProvider,
} from '../communications'

export class FakeTwilioProvider
  implements
    PhoneVerificationProvider,
    PhoneLookupProvider,
    MessagingProvider,
    VoiceProvider
{
  async createVerification(params: {
    toNumber: string
    channel: string
  }): Promise<PhoneVerification> {
    return {
      ...EMPTY_VERIFICATION_FIELDS,
      provider: 'fake',
      providerSid: this.sid('verify', params.toNumber, params.channel),
      status: 'pending',
      toNumber: params.toNumber,
      channel: params.channel,
    }
  }

  async approveVerification(params: {
    toNumber: string
    code: string
  }): Promise<PhoneVerification> {
    const approved = params.code === '000000'
    return {
      ...EMPTY_VERIFICATION_FIELDS,
      provider: 'fake',
      providerSid: this.sid('check', params.toNumber),
      status: approved ? 'approved' : 'pending',
      toNumber: params.toNumber,
      channel: '',
      valid: approved,
    }
  }

  async createLookup(params: {
    number: string
    includeLineType?: boolean
  }): Promise<PhoneLookup> {
    return {
      provider: 'fake',
      number: params.number,
      nationalFormat: null,
      countryCode: null,
      valid: true,
      carrierName: null,
      lineType: null,
      mobileCountryCode: null,
      mobileNetworkCode: null,
    }
  }

  async createMessage(params: {
    toNumber: string
    body: string | null
    channel: string
    contentSid?: string | null
    statusCallback?: string | null
  }): Promise<ProviderMessage> {
    return {
      provider: 'fake',
      providerSid: this.sid('message', params.toNumber, params.channel),
      status: 'queued',
      toNumber: params.toNumber,
      fromNumber: null,
    }
  }

  async retrieveMessage(params: {
    providerSid: string
  }): Promise<ProviderMessage> {
    return {
      provider: 'fake',
      providerSid: params.providerSid,
      status: 'queued',
      toNumber: '',
      fromNumber: null,
    }
  }

  async createCall(params: {
    toNumber: string
    twimlUrl: string
    statusCallback?: string | null
  }): Promise<ProviderCall> {
    return {
      provider: 'fake',
      providerSid: this.sid('call', params.toNumber),
      status: 'queued',
      toNumber: params.toNumber,
      fromNumber: null,
    }
  }

  async retrieveCall(params: { providerSid: string }): Promise<ProviderCall> {
    return {
      provider: 'fake',
      providerSid: params.providerSid,
      status: 'queued',
      toNumber: '',
      fromNumber: null,
    }
  }

  private sid(...parts: string[]): string {
    return (
      'fake_' +
      createHash('sha256').update(parts.join('|')).digest('hex').slice(0, 24)
    )
  }
}
